%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: camera_info_broadcaster
components:
- type: nvidia::gxf::Broadcast
  parameters:
    source: camera_info_in
- name: camera_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_info_in
    min_size: 1
- name: camera_info_out_1
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_info_out_1
    min_size: 1
- name: camera_info_out_2
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_info_out_2
    min_size: 1
---
name: cuda_april_tag_component
components:
- name: camera_image_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- name: april_tags_out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- name: allocator
  type: nvidia::gxf::UnboundedAllocator
- name: cuda_april_tag_detector
  type: nvidia::isaac::CudaAprilTagDetector
  parameters:
    camera_image: camera_image_info_in
    april_tags: april_tags_out
    allocator: allocator
    max_tags: 64
    tag_dimensions: 0.22
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_image_info_in
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: april_tags_out
    min_size: 1
---
name: resource
components:
- name: adapter
  type: nvidia::cvcore::tensor_ops::ImageAdapter
  parameters:
    message_type: "VideoBuffer"
---
name: image_converter
components:
- name: camera_image_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_image_in
    min_size: 1
- name: camera_image_out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_image_out
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 14400000
    num_blocks: 40
- name: image_converter
  type: nvidia::cvcore::tensor_ops::ConvertColorFormat
  parameters:
    output_type: "RGB_U8"
    receiver: camera_image_in
    transmitter: camera_image_out
    pool: pool
    input_adapter: resource/adapter
    output_adapter: resource/adapter
    output_name: "image"
---
name: camera_image_info_compositor
components:
- name: camera_image_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_image_in
    min_size: 1

- name: camera_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_info_in
    min_size: 1
  
- name: camera_image_info_out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_image_info_out
    min_size: 1

- name: camera_image_info_compositor
  type: nvidia::isaac_ros::MessageCompositor
  parameters:
    output_image_name: frame
    image_rx: camera_image_in
    camera_model_rx: camera_info_in
    tx: camera_image_info_out
---
name: timestamp_compositor
components:
- name: april_tags_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: april_tags_in
    min_size: 1

- name: camera_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_info_in
    min_size: 1
  
- name: april_tags_timestamp_out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: april_tags_timestamp_out
    min_size: 1

- name: timestamp_compositor
  type: nvidia::isaac_ros::MessageCompositor
  parameters:
    image_rx: april_tags_in
    camera_model_rx: camera_info_in
    tx: april_tags_timestamp_out
---
name: vault
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: vault
  type: nvidia::gxf::Vault
  parameters:
    source: signal
    max_waiting_count: 1
    drop_waiting: false
---
components:
- name: edge0
  type: nvidia::gxf::Connection
  parameters:
    source: camera_image_info_compositor/camera_image_info_out
    target: cuda_april_tag_component/camera_image_info_in
- name: edge1
  type: nvidia::gxf::Connection
  parameters:
    source: camera_info_broadcaster/camera_info_out_1
    target: camera_image_info_compositor/camera_info_in
- name: edge2
  type: nvidia::gxf::Connection
  parameters:
    source: camera_info_broadcaster/camera_info_out_2
    target: timestamp_compositor/camera_info_in
- name: edge3
  type: nvidia::gxf::Connection
  parameters:
    source: cuda_april_tag_component/april_tags_out
    target: timestamp_compositor/april_tags_in
- name: edge4
  type: nvidia::gxf::Connection
  parameters:
    source: timestamp_compositor/april_tags_timestamp_out
    target: vault/signal
- name: edge5
  type: nvidia::gxf::Connection
  parameters:
    source: image_converter/camera_image_out
    target: camera_image_info_compositor/camera_image_in
---
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::MultiThreadScheduler
  parameters:
    clock: clock
    worker_thread_number: 2
    stop_on_deadlock: false
- type: nvidia::gxf::JobStatistics
  parameters:
    clock: clock
